README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# SNPs Enrichment Analysis With Sampling <!--<img src="man/figures/snpEnrichment_hex.png" align="right" width="120" /> -->

<!-- badges: start -->

[![Lifecycle:
retired](https://img.shields.io/badge/lifecycle-retired-orange.svg)](https://www.tidyverse.org/lifecycle/#retired)
[![GitHub
tag](https://img.shields.io/github/tag/mcanouil/snpEnrichment.svg?label=%22latest%20tag%22)](https://github.com/mcanouil/snpEnrichment)
[![Travis-CI Build
Status](https://travis-ci.org/mcanouil/snpEnrichment.svg?branch=master)](https://travis-ci.org/mcanouil/snpEnrichment)
[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/mcanouil/snpEnrichment?branch=master&svg=true)](https://ci.appveyor.com/project/mcanouil/snpEnrichment)
[![Coverage Status
(codecov)](https://codecov.io/gh/mcanouil/snpEnrichment/branch/master/graph/badge.svg)](https://codecov.io/gh/mcanouil/snpEnrichment)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version-ago/snpEnrichment)](https://cran.r-project.org/package=snpEnrichment)
[![cran
checks\_worst](https://cranchecks.info/badges/worst/snpEnrichment)](https://cran.r-project.org/web/checks/check_results_snpEnrichment.html)
[![CRAN\_Download\_total](http://cranlogs.r-pkg.org/badges/grand-total/snpEnrichment)](https://cran.r-project.org/package=snpEnrichment)
<!--[![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/2719/badge)](https://bestpractices.coreinfrastructure.org/projects/2719) -->
<!--[![cran checks_summary](https://cranchecks.info/badges/summary/snpEnrichment)](https://cran.r-project.org/web/checks/check_results_snpEnrichment.html) -->
<!--[![CRAN_Download_month](http://cranlogs.r-pkg.org/badges/snpEnrichment?color=brightgreen)](https://cran.r-project.org/package=snpEnrichment) -->
<!--[![Coverage Status (coveralls)](https://coveralls.io/repos/github/mcanouil/snpEnrichment/badge.svg?branch=master)](https://coveralls.io/github/mcanouil/snpEnrichment?branch=master) -->
<!-- badges: end -->

Implements classes and methods for large scale SNP enrichment analysis
(*e.g.*, SNPs associated with genes expression in a GWAS signal).

**Not maintained anymore.**

## Installation

``` r
# Install snpEnrichment from CRAN:
install.packages("snpEnrichment")

# Or the the development version from GitHub:
# install.packages("remotes")
remotes::install_github("mcanouil/snpEnrichment")
```

## Overview

**Note**: Internal data management in ‘snpEnrichment’ use RefSNP (rs)
IDs.

### Load snpEnrichment

``` r
library(snpEnrichment)
```

### Prepare data

``` r
snpInfoDir <- system.file("extdata/snpInfo", package = "snpEnrichment")
signalFile <- system.file("extdata/Signal/toySignal.txt", package = "snpEnrichment")

initFiles(pattern = "Chrom", snpInfoDir, signalFile, mc.cores = 1)

writeLD(
  pattern = "Chrom", 
  snpInfoDir, 
  signalFile, 
  ldDir = NULL,
  ldThresh = 0.8, 
  depth = 1000, 
  mc.cores = 1
)
```

### Read data

``` r
snpListDir <- system.file("extdata/List", package = "snpEnrichment")
data(transcript)
transcriptFile <- transcript

toyData <- readEnrichment(
    pattern = "Chrom", 
    signalFile, 
    transcriptFile,
    snpListDir, 
    snpInfoDir,
    distThresh = 1000, 
    sigThresh = 0.05, 
    LD = TRUE, 
    ldDir = NULL,
    mc.cores = 1
)
toyData
```

### Compute results

``` r
reSample(
  object = toyData,
  nSample = 10,
  empiricPvalue = TRUE,
  MAFpool = c(0.05, 0.10, 0.2, 0.3, 0.4, 0.5),
  mc.cores = 1,
  onlyGenome = TRUE
)
```

### 5\. Further analysis: Exclude SNP from original list.

``` r
excludeFile <- c(
  "rs4376885", "rs17690928", "rs6460708", "rs13061537", "rs11769827",
  "rs12717054", "rs2907627", "rs1380109", "rs7024214", "rs7711972",
  "rs9658282", "rs11750720", "rs1793268", "rs774568", "rs6921786",
  "rs1699031", "rs6994771", "rs16926670", "rs465612", "rs3012084",
  "rs354850", "rs12803455", "rs13384873", "rs4364668", "rs8181047",
  "rs2179993", "rs12049335", "rs6079926", "rs2175144", "rs11564427",
  "rs7786389", "rs7005565", "rs17423335", "rs12474102", "rs191314",
  "rs10513168", "rs1711437", "rs1992620", "rs283115", "rs10754563",
  "rs10851727", "rs2173191", "rs7661353", "rs1342113", "rs7042073",
  "rs1567445", "rs10120375", "rs550060", "rs3761218", "rs4512977"
)

# or
excludeFile <- system.file("extdata/Exclude/toyExclude.txt", package = "snpEnrichment")
```

``` r
toyData_exclude <- excludeSNP(toyData, excludeFile, mc.cores = 1)
```

**Warning:** `compareEnrichment` is in development\!\!

``` r
compareResults <- compareEnrichment(
  object.x = toyData,
  object.y = toyData_exclude,
  pattern = "Chrom",
  nSample = 10,
  empiricPvalue = TRUE,
  mc.cores = 1,
  onlyGenome = TRUE
)
```

### Watch results

``` r
show(toyData)
print(toyData)
head(getEnrichSNP(toyData, type = "xSNP"))

show(toyData_exclude)
print(toyData_exclude)
head(getEnrichSNP(toyData_exclude, type = "eSNP"))
```

## Getting help

If you encounter a clear bug, please file a minimal reproducible example
on [github](https://github.com/mcanouil/CARoT/issues). For questions and
other discussion, please contact the package maintainer.

-----

Please note that this project is released with a [Contributor Code of
Conduct](.github/CODE_OF_CONDUCT.md). By participating in this project
you agree to abide by its terms.


NEWS
2016-02-09 Mickael Canouil
        * DESCRIPTION: version is 1.7.2
        * R/*.R: Minor code optimisation.

2015-10-01 Mickael Canouil
        * DESCRIPTION: version is 1.7.1
        * R/*.R: Minor code optimisation.

2015-10-01 Mickael Canouil
        * DESCRIPTION: version is 1.7.0
        * NAMESPACE: Update to fix warnings from R CMD check using R version 3.2.2.
        * DESCRIPTION: Update according to new imports in NAMESPACE.

2015-10-01 Mickael Canouil
        * DESCRIPTION: version is 1.6.6
        * NAMESPACE: Update to fix warnings from R CMD check using R version 3.2.2.
        * R/enrichment.R: Update 'plot' method to fix global definition warnings.

2015-04-30 Mickael Canouil
        * DESCRIPTION: version is 1.6.5
        * tests/RUN-TEST_snpEnrichment.R: Add a test script.

2015-03-10 Mickael Canouil
        * DESCRIPTION: version is 1.6.4
        * R/enrichment.R: Update 'getEnrichSNP' to add information about related eSNP for xSNP.

2015-02-09 Mickael Canouil
        * DESCRIPTION: version is 1.6.3
        * DESCRIPTION: Change maintainer email.
        * R/enrichment.R: Minor changes.

2015-01-05 Mickael Canouil
        * DESCRIPTION: version is 1.6.2
        * R/enrichment.R: Minor changes.
        * R/chromosome.R: Minor changes.
        * R/Global.R: Minor changes.

2014-11-17 Mickael Canouil
        * DESCRIPTION: version is 1.6.1
        * R/enrichment.R: Minor changes.

2014-11-04 Mickael Canouil
        * DESCRIPTION: version is 1.6.0
        * R/Global.R: Fix a bug in 'compareEnrichment' method with parameter 'empiricPvalue=TRUE'.


2014-11-03 Mickael Canouil
        * DESCRIPTION: version is 1.5.10
        * R/Global.R: Fix a glitch in 'readEnrichment'. Sometimes chromosomes 10 and 20 are read respectively instead of 1 and 2.
        * R/Global.R: Some minor changes.


2014-10-14 Mickael Canouil
        * DESCRIPTION: version is 1.5.9
        * data/toyEnrichment.RData: Minor changes.
        * R/Global.R: Add internal function '.splitByChrom'.
            User is now allowed to use a single file in 'snpListDir' for 'readEnrichment' function.
        * man/snpEnrichment-internal.Rd: Update according to R code changes.
        * man/readEnrichment.Rd: Update according to R code changes.


2014-10-14 Mickael Canouil
        * DESCRIPTION: version is 1.5.8
        * R/enrichment.R: Update 'print' method to print results according to the parameter 'empiricPvalue' in 'reSample' function.
        * R/enrichment.R: Update 'compareEnrichment' according to changes in 'print' method.
        * R/chromosome.R: Some minor changes.
        * man/compareEnrichment.Rd: Update according to R code changes.
        * man/print-methods.Rd: Update according to R code changes.


2014-10-13 Mickael Canouil
        * DESCRIPTION: version is 1.5.7
        * R/*.R: Add function 'getEnrichSNP' and some minor corrections.
        * man/*.Rd: Update according to R code changes and some corrections.
        * NAMESPACE: Updata according to R code changes.
        * inst/example.R: Update 'excludeFile' list according to changes in 'toyEnrichment' and 'extdata'.
        * man/snpEnrichment-package.Rd: Update 'excludeFile' list according to changes in 'toyEnrichment' and 'extdata'.
        * man/excludeSNP-methods.Rd: Update 'excludeFile' list according to changes in 'toyEnrichment' and 'extdata'.
        * README.md: Update 'excludeFile' list according to changes in 'toyEnrichment' and 'extdata'.


2014-10-10 Mickael Canouil
        * DESCRIPTION: version is 1.5.6
        * R/Global.R: 'readEnrichment' works with at least one snp list in 'snpListDir'.
        * R/Global.R: Fix 'mclapply2' under operating systems which are not Linux.
        * R/Global.R: Fix 'maxCores' under operating systems which are not Linux.
        * man/snpEnrichment-package.Rd: Update example.
        * inst/example.R: Update according to changes in 'snpEnrichment-package.Rd'.
        * README.md: Update according to changes in 'snpEnrichment-package.Rd'.
        * R/*.R: the "types" parameter is renamed "type".
        * man/*.Rd: Update according to R code changes.


2014-10-10 Mickael Canouil
        * DESCRIPTION: version is 1.5.5
        * man/*.Rd: Replace 'summary' by 'print'.
        * inst/example.R: Replace 'summary' by 'print' according to changes in '*.Rd' files.
        * README.md: Replace 'summary' by 'print' according to changes in '*.Rd' files.


2014-10-10 Mickael Canouil
        * DESCRIPTION: version is 1.5.4
        * R/*.R: Change default value for 'empiricPvalue' parameter to 'TRUE'.
        * man/*.Rd: Update according to R code changes.
        * inst/example.R: Update according to R code changes.
        * README.md: Update according to R code changes.
        * R/enrichment.R: Change warning messages in 'compareEnrichment'.


2014-10-10 Mickael Canouil
        * DESCRIPTION: version is 1.5.3
        * R/Global.R: Fix an issue under Windows systems with mclapply
        * R/Global.R: Update warning message in 'reSample' method.
        * R/Global.R: Fix an issue in 'initFiles', 'writeLD' and 'readEnrichment'.
            Some bim files could not found according to the sorting method in the folder.
        * R/enrichment.R: 'plot' method is consitent with parameters used in 'reSample' (e.g. 'empiricPvalue')
        * README.md: Some corrections.
        * R/*.R: Change default value for 'onlyGenome' parameter to 'TRUE'.
        * man/*.Rd: Update according to R code changes.


2014-09-29 Mickael Canouil
        * DESCRIPTION: version is 1.5.2
        * R/Global.R: Add some internal functions to check files needed by 'initFiles', 'writeLD' and 'readEnrichment' functions.
        * man/snpEnrichment-internal.Rd: Update according to R code changes.
        * inst/example.R: Update according to R code changes.


2014-09-29 Mickael Canouil
        * DESCRIPTION: version is 1.5.1
        * R/Global.R: Internal function 'maxCores' handle warnings on macintosh system.

R/plot.R
#' Plot method (S4) for [Enrichment-class] object
#'
#' [plot] is a generic function for plotting of R objects.
#' The function invokes particular [methods] which depend on the [class] of the first argument.
#'
#' @param x An object of class [Enrichment-class] which the Z statistics or p-values have to be drawn.
#' @param what Default `"Genome"`. Plot Z statistics or p-values for the defined `what`,
#'   *i.e.*, `"All"`, `"Genome"` or numeric vector.
#' @param type Plot the selected analysis for `"eSNP"` and/or `"xSNP"`.
#' @param ggplot Use ggplot (default `FALSE`) instead of classic plot method.
#' @param pvalue If `TRUE`, p-value convergense is plotted. Otherwise, Z statistic is plotted.
#' @param ... Arguments to be passed to methods, such as graphical parameters (see [par]).
#'
#' @examples
#' if (interactive()) {
#'   data(toyEnrichment)
#'   reSample(toyEnrichment, 10)
#'   plot(toyEnrichment)
#' }
#'
#' @name plot
#' @exportMethod plot
methods::setGeneric(name = "plot", def = function (x, y, ...) standardGeneric("plot"))
#' @name plot
#' @rdname plot
#' @aliases plot,Enrichment-method
methods::setMethod(f = "plot", signature = "Enrichment", definition = function(x, what = "Genome", type = c("eSNP", "xSNP"), ggplot = FALSE, pvalue = TRUE, ...) {
  if (is.null(unlist(x@Call$reSample, use.names = FALSE))) {
    stop('[Enrichment:plot] "reSample" have to be run before "plot".', call. = FALSE)
  }

  if (is.null(what) | any(!what %in% c("All", "Genome", seq_len(22)))) {
    stop('[Enrichment:plot] "what" must be: "All", "Genome" or numeric value (atomic or vector).', call. = FALSE)
  }
  if (is.null(type) | any(!type %in% c("eSNP", "xSNP"))) {
    stop('[Enrichment:plot] "type" must be: "eSNP" and/or "xSNP".', call. = FALSE)
  }
  if (any(type %in% "xSNP") & length(x["xSNP"]["List"]) == 0) {
    type <- "eSNP"
  }

  .computeER4plot <- function(EnrichSNPObject) {
    ER <- EnrichSNPObject@EnrichmentRatio
    if (nrow(EnrichSNPObject@Resampling) == 0) {
      stop('[Enrichment:plot] "reSample" have to be run before "plot".', call. = FALSE)
    }
    resampling <- EnrichSNPObject@Resampling[, 5]
    ERsample <- NULL
    size <- length(resampling)
    if (size >= 1000) {
      interv <- unique(c(seq(from = min(1000, floor(0.1 * size)), to = size, by = floor(size / 1000)), size))
    } else {
      interv <- unique(c(seq(from = max(floor(0.1 * size), 3), to = size, by = 1), size))
    }
    ERsample <- sapply(interv, function(k) {
      resamplingInterv <- resampling[1:k]
      resamplingClean <- resamplingInterv[!(is.na(resamplingInterv) | is.infinite(resamplingInterv))]
      mu <- mean(resamplingClean)
      sigma <- sqrt(stats::var(resamplingClean))

      if (sigma == 0 | is.na(sigma)) {
        if (mu == 0) 0 else (ER - mu)
      } else {
        (ER - mu) / sigma
      }
    })
    names(ERsample) <- interv
    as.matrix(ERsample)
  }
  .computeEmpP4plot <- function(EnrichSNPObject) {
    ER <- EnrichSNPObject@EnrichmentRatio
    if (nrow(EnrichSNPObject@Resampling) == 0) {
      stop('[Enrichment:plot] "reSample" have to be run before "plot".', call. = FALSE)
    }
    resampling <- EnrichSNPObject@Resampling[, 5]
    ERsample <- NULL
    size <- length(resampling)
    if (size >= 1000) {
      interv <- unique(c(seq(from = min(1000, floor(0.1 * size)), to = size, by = floor(size / 1000)), size))
    } else {
      interv <- unique(c(seq(from = max(floor(0.1 * size), 3), to = size, by = 1), size))
    }
    ERsample <- sapply(interv, function(k) {
      resamplingInterv <- resampling[1:k]
      resamplingClean <- resamplingInterv[!(is.na(resamplingInterv) | is.infinite(resamplingInterv))]
      sum(EnrichSNPObject@EnrichmentRatio < resamplingClean) / length(resamplingClean)
    })
    names(ERsample) <- interv
    as.matrix(ERsample)
  }

  if (x@Call$reSample$empiricPvalue) {
    matrixER <- list(eSNP = NULL, xSNP = NULL)
    for (iType in type) {
      if (length(what) == 1) {
        switch(EXPR = as.character(what),
          "Genome" = {
            matrixER[[iType]] <- .computeEmpP4plot(x[iType])
            colnames(matrixER[[iType]]) <- "Genome"
          },
          "All" = {
            matrixER[[iType]] <- .computeEmpP4plot(x[iType])
            for (j in seq_len(22)) {
              matrixER[[iType]] <- cbind(matrixER[[iType]], .computeEmpP4plot(x["Chromosomes", j][iType]))
            }
            colnames(matrixER[[iType]]) <- c("Genome", paste0("Chrom", seq_len(22)))
          }, {
            for (j in what) {
              matrixER[[iType]] <- cbind(matrixER[[iType]], .computeEmpP4plot(x["Chromosomes", j][iType]))
            }
            colnames(matrixER[[iType]]) <- paste0("Chrom", what)
          }
        )
      } else {
        for (j in what) {
          matrixER[[iType]] <- cbind(matrixER[[iType]], .computeEmpP4plot(x["Chromosomes", j][iType]))
        }
        colnames(matrixER[[iType]]) <- paste0("Chrom", what)
      }
    }
  } else {
    matrixER <- list(eSNP = NULL, xSNP = NULL)
    for (iType in type) {
      if (length(what) == 1) {
        switch(EXPR = as.character(what),
          "Genome" = {
            matrixER[[iType]] <- .computeER4plot(x[iType])
            colnames(matrixER[[iType]]) <- "Genome"
          },
          "All" = {
            matrixER[[iType]] <- .computeER4plot(x[iType])
            for (j in seq_len(22)) {
              matrixER[[iType]] <- cbind(matrixER[[iType]], .computeER4plot(x["Chromosomes", j][iType]))
            }
            colnames(matrixER[[iType]]) <- c("Genome", paste0("Chrom", seq_len(22)))
          }, {
            for (j in what) {
              matrixER[[iType]] <- cbind(matrixER[[iType]], .computeER4plot(x["Chromosomes", j][iType]))
            }
            colnames(matrixER[[iType]]) <- paste0("Chrom", what)
          }
        )
      } else {
        for (j in what) {
          matrixER[[iType]] <- cbind(matrixER[[iType]], .computeER4plot(x["Chromosomes", j][iType]))
        }
        colnames(matrixER[[iType]]) <- paste0("Chrom", what)
      }
    }
  }
  if (ggplot) {
    if (!require("ggplot2") | !require("grid")) {
      stop('[Enrichment:plot] "ggPlot2" and "grid" packages must be installed with "ggplot=TRUE".', call. = FALSE)
    }
    multiplot <- function(..., plotlist = NULL, cols = 1, rows = 1, layout = NULL) {
      plots <- c(list(...), plotlist)
      numPlots <- length(plots)
      if (is.null(layout)) {
        layout <- matrix(seq_len(cols * ceiling(numPlots / cols)), ncol = cols, nrow = rows, byrow = TRUE)
      }
      if (numPlots == 1) {
        print(plots[[1]])
      } else {
        grid::grid.newpage()
        grid::pushViewport(grid::viewport(layout = grid::grid.layout(nrow(layout), ncol(layout))))
        for (i in 1:numPlots) {
          matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))
          print(plots[[i]], vp = grid::viewport(layout.pos.row = matchidx$row, layout.pos.col = matchidx$col))
        }
      }
    }
    .ggplotColours <- function(n = 6, h = c(0, 360) + 15) {
      if ((diff(h) %% 360) < 1) {
        h[2] <- h[2] - 360 / n
      }
      grDevices::hcl(h = (seq(h[1], h[2], length = n)), c = 100, l = 65)
    }
    listPlots <- list()
    for (iType in type) {
      if (pvalue) {
        if (x@Call$reSample$empiricPvalue) {
          ylab <- "P-Value (Empirical)"
        } else {
          matrixER[[iType]] <- apply(matrixER[[iType]], 2, stats::pnorm, lower.tail = FALSE)
          ylab <- "P-Value (From Z-statistic)"
        }
      } else {
        ylab <- "Z statistic"
      }
      if (ncol(matrixER[[iType]]) > 1) {
        matrixER[[iType]] <- apply(matrixER[[iType]], 2, scale)
      }
      tmpDF <- as.data.frame(t(matrixER[[iType]]))
      cnames <- colnames(tmpDF)
      colnames(tmpDF) <- paste0("R", colnames(tmpDF))
      tmpDF$IID <- factor(colnames(matrixER[[iType]]), levels = c("Genome", paste0("Chrom", seq_len(22))), labels = c("Genome", paste0("Chrom", seq_len(22))))
      tmp <- stats::reshape(tmpDF, idvar = "IID", direction = "long", varying = list(grep("R", colnames(tmpDF))), times = cnames)
      colnames(tmp) <- c("IID", "Resampling", "Z")
      tmp[, "Resampling"] <- as.numeric(tmp[, "Resampling"])

      p <- ggplot2::ggplot(data = tmp, ggplot2::aes_string(x = "Resampling", y = "Z", colour = "IID")) + ggplot2::geom_line()
      noGridColour <- "transparent"
      base_size <- 12
      base_family <- ""
      p <- p + ggplot2::theme(
        line = ggplot2::element_line(colour = "black", size = 0.5, linetype = 1, lineend = "butt"),
        rect = ggplot2::element_rect(fill = "white", colour = "black", size = 0.5, linetype = 1),
        text = ggplot2::element_text(family = base_family, face = "plain", colour = "black", size = base_size, hjust = 0.5, vjust = 0.5, angle = 0, lineheight = 0.9),
        axis.text = ggplot2::element_text(size = ggplot2::rel(0.8), colour = "black"),
        strip.text = ggplot2::element_text(size = ggplot2::rel(0.8)),
        axis.line = ggplot2::element_blank(),
        axis.text.x = ggplot2::element_text(vjust = 1),
        axis.text.y = ggplot2::element_text(hjust = 1),
        axis.ticks = ggplot2::element_line(colour = "black"),
        axis.title.x = ggplot2::element_text(),
        axis.title.y = ggplot2::element_text(angle = 90),
        axis.ticks.length = ggplot2::unit(0.15, "cm"),
        axis.ticks.margin = ggplot2::unit(0.1, "cm"),
        legend.background = ggplot2::element_rect(fill = "white", colour = "black"),
        legend.margin = ggplot2::unit(0.2, "cm"),
        legend.key = ggplot2::element_rect(fill = "white", colour = "black"),
        legend.key.size = ggplot2::unit(1.2, "lines"),
        legend.key.height = NULL,
        legend.key.width = NULL,
        legend.text = ggplot2::element_text(size = ggplot2::rel(0.8)),
        legend.text.align = NULL,
        legend.title = ggplot2::element_text(size = ggplot2::rel(0.8), face = "bold", hjust = 0),
        legend.title.align = NULL,
        legend.position = "right",
        legend.direction = NULL,
        legend.justification = "center",
        legend.box = NULL,
        panel.background = ggplot2::element_rect(fill = "white", colour = "black"),
        panel.border = ggplot2::element_blank(),
        panel.grid.major = ggplot2::element_line(colour = noGridColour[1]),
        panel.grid.minor = ggplot2::element_line(colour = noGridColour[length(noGridColour)], size = 0.25),
        panel.margin = ggplot2::unit(0.25, "lines"),
        strip.background = ggplot2::element_rect(fill = "black", colour = "black"),
        strip.text.x = ggplot2::element_text(colour = "white"),
        strip.text.y = ggplot2::element_text(angle = -90, colour = "white"),
        plot.background = ggplot2::element_rect(colour = "white"),
        plot.title = ggplot2::element_text(size = ggplot2::rel(1.2)),
        plot.margin = ggplot2::unit(c(1, 1, 0.5, 0.5), "lines"),
        complete = TRUE
      )
      p <- p + ggplot2::xlab(paste(iType, "Resampling"))
      if (ncol(matrixER[[iType]]) > 1) {
        p <- p + ggplot2::ylab(paste(ylab, "(scale and center)"))
      } else {
        p <- p + ggplot2::ylab(ylab)
      }
      p <- p + ggplot2::theme(legend.title = ggplot2::element_blank())
      if (length(what) >= 5 | what == "All") {
        p <- p + ggplot2::theme(legend.background = ggplot2::element_rect(fill = "gray90", linetype = "dotted"))
      } else {
        quarter <- floor(3 / 4 * nrow(tmp)):nrow(tmp)
        rangeZtmp <- range(tmp[, "Z"])
        if (rangeZtmp[1] == rangeZtmp[2]) {
          if (all(rangeZtmp == 0)) {
            rangeZtmp[1] <- -1
            rangeZtmp[2] <- 1
          } else {
            rangeZtmp[1] <- rangeZtmp[1] * 0.90
            rangeZtmp[2] <- rangeZtmp[2] * 1.10
          }
        }
        rangeZ <- seq(rangeZtmp[1], rangeZtmp[2], by = diff(rangeZtmp) * 1 / 3)
        names(rangeZ) <- c("0%", "33%", "66%", "100%")
        rangeQuarter <- range(tmp[quarter, "Z"])
        inf <- apply(sapply(rangeQuarter, function(lim) lim < rangeZ), 1, all)
        sup <- apply(sapply(rangeQuarter, function(lim) lim > rangeZ), 1, all)
        if (sum(inf) <= sum(sup)) {
          p <- p + ggplot2::theme(legend.justification = c(1, 0), legend.position = c(1, 0))
        } else {
          p <- p + ggplot2::theme(legend.justification = c(1, 1), legend.position = c(1, 1))
        }
      }

      if ("Genome" %in% unique(tmp$IID)) {
        p <- p + ggplot2::scale_colour_manual(values = c("black", .ggplotColours(ifelse(length(unique(tmp$IID)) - 1 > 0, length(unique(tmp$IID)) - 1, 1))))
      }
      listPlots[[iType]] <- p
    }
    multiplot(plotlist = listPlots, cols = length(listPlots))
    invisible(listPlots)
  } else {
    graphics::par(mfrow = c(1, length(type)))
    for (iType in type) {
      if (pvalue) {
        if (x@Call$reSample$empiricPvalue) {
          matrixER[[iType]] <- .computeEmpP4plot(x[iType])
          ylab <- "P-Value (Empirical)"
        } else {
          matrixER[[iType]] <- apply(matrixER[[iType]], 2, stats::pnorm, lower.tail = FALSE)
          ylab <- "P-Value (From Z-statistic)"
        }
      } else {
        ylab <- "Z statistic"
      }
      nbCol <- ncol(matrixER[[iType]])
      ylim <- range(stats::na.exclude(matrixER[[iType]]))
      xNames <- rownames(matrixER[[iType]])
      colors <- grDevices::rainbow(nbCol)
      if (nbCol > 1) {
        matrixER[[iType]] <- apply(matrixER[[iType]], 2, scale)
        ylab <- paste(ylab, "(scale and center)")
        graphics::plot(x = xNames, y = matrixER[[iType]][, 1], ylab = ylab, xlab = iType, type = "l", ylim = ylim)
        res <- sapply(seq_len(ncol(matrixER[[iType]][, -1])), function(iER) {
          graphics::lines(x = xNames, y = matrixER[[iType]][, iER + 1], iType = "l", ylim = ylim, col = colors[iER + 1])
        })
      } else {
        graphics::plot(x = xNames, y = matrixER[[iType]][, 1], ylab = ylab, xlab = iType, type = "l", ylim = ylim)
      }
    }
    invisible()
  }
})
